### PR TITLE
Snesmf reuse base sanity check forward compatability

### PR DESCRIPTION
**Pull request by lindsayad**
libmesh/libmesh#2646 will provided the sanity check asked for in #14644. As that PR shows, we alraedy have some problems that require setting `snesmf_reuse_base=false`. We need this to be merged first so that libmesh/libmesh#2646 can be green and we don't lose CI.

This also adds `-snes_converged_reason` and `-ksp_converged_reason` to all simulations unless a user explicitly asks for it to be turned off in the output block (similar to how linear residual printing can be turned off)

### CODE REVIEW

**Comment by lindsayad** [comment]
@fdkong I think this is kind of cool. We can do some detection of when it is unsafe for people to reuse the nonlinear residual for the initial linear residual.

**Comment by fdkong** [comment]
Yes, it is cool. In the same PR, you could also do that non-converged prints we talked earlier.

**Comment by lindsayad** [comment]
I guess that the vector I get from `KSPGetRHS` is safe to compare against the nonlinear residual, even for left-side preconditioning...?

**Comment by fdkong** [comment]
> I guess that the vector I get from `KSPGetRHS` is safe to compare against the nonlinear residual, even for left-side preconditioning...? E.g. I am getting, b, and not P_inverse * b

I think so.


```
1525: PetscErrorCode  KSPGetRhs(KSP ksp,Vec *r)
1526: {
1530:   *r = ksp->vec_rhs;
1531:   return(0);
1532: }
```

**Comment by lindsayad** [comment]
Does it make sense in https://civet.inl.gov/job/561047/, that the residual evaluation should be stateful (e.g. not just a function of the current solution) for the `singular_contaminated` test but not for the `singular` test? I'm not sure that it does just looking briefly at the objects involved. @fdkong it looks like you've touched these tests but the original owner is @snschune

**Comment by lindsayad** [comment]
I need a sanity check on my sanity check

**Comment by fdkong** [comment]
> Does it make sense in https://civet.inl.gov/job/561047/, that the residual evaluation should be stateful (e.g. not just a function of the current solution) for the `singular_contaminated` test but not for the `singular` test? I'm not sure that it does just looking briefly at the objects involved. @fdkong it looks like you've touched these tests but the original owner is @snschune

You should skip tests handling null spaces.

**Comment by lindsayad** [comment]
> You should skip tests handling null spaces.

why?

**Comment by lindsayad** [comment]
I think that my bitwise comparison probably is not safe in parallel. I'll need to restrict that sanity check to serial runs I think

**Comment by fdkong** [comment]
> why?

`rhs` is changed because of  removing transpose null spaces

**Comment by lindsayad** [comment]
Beautiful catching of contact statefulness if I try to run `reuse_base` on:
```
test:verification/patch_tests/plane_1.glued_kin ........................................................... OK
test:mortar_tm.horizontal_blocks_mortar_TM ................................................................ OK
test:verification/patch_tests/ring_1.glued_kin ............................................................ OK
test:verification/patch_tests/cyl_4.glued_kin ............................................................. OK
test:verification/patch_tests/cyl_2.glued_kin ............................................................. OK
test:verification/patch_tests/ring_2.glued_kin ............................................................ OK
test:verification/patch_tests/cyl_3.glued_kin ............................................................. OK
test:verification/patch_tests/single_pnt_2d.glued_kin ..................................................... OK
test:verification/patch_tests/plane_2.glued_kin ........................................................... OK
test:verification/patch_tests/plane_3.glued_kin ........................................................... OK
test:verification/patch_tests/ring_3.glued_kin ............................................................ OK
test:verification/patch_tests/cyl_1.glued_kin ............................................................. OK
test:verification/patch_tests/plane_4.glued_kin ........................................................... OK
test:verification/patch_tests/ring_4.glued_kin ............................................................ OK
test:verification/patch_tests/brick_1.glued_kin ........................................................... OK
test:normalized_penalty.normalized_penalty ................................................................ OK
test:verification/patch_tests/automatic_patch_update.error_test .................... [skipped dependency] SKIP
test:check_error.contact_displacements .................................................................... OK
test:nodal_area.3D ........................................................................................ OK
test:verification/patch_tests/brick_3.frictionless_kin .................................................... OK
test:mechanical-small-problem.mortar-mech ................................................................. OK
test:verification/patch_tests/brick_2.glued_kin ........................................................... OK
test:verification/patch_tests/plane_1.glued_pen ........................................................... OK
test:verification/patch_tests/ring_1.glued_pen ............................................................ OK
test:verification/patch_tests/cyl_2.glued_pen ............................................................. OK
test:verification/patch_tests/cyl_4.glued_pen ............................................................. OK
test:verification/patch_tests/ring_2.glued_pen ............................................................ OK
test:verification/patch_tests/plane_2.glued_pen ........................................................... OK
test:verification/patch_tests/cyl_3.glued_pen ............................................................. OK
test:verification/patch_tests/single_pnt_2d.glued_pen ..................................................... OK
test:verification/patch_tests/plane_3.glued_pen ........................................................... OK
test:verification/patch_tests/ring_3.glued_pen ............................................................ OK
test:verification/patch_tests/cyl_1.glued_pen ............................................................. OK
test:verification/patch_tests/brick_1.glued_pen ........................................................... OK
test:verification/patch_tests/ring_1.frictionless_kin ..................................................... OK
test:verification/patch_tests/plane_1.frictionless_kin .................................................... OK
test:verification/patch_tests/cyl_2.frictionless_kin ...................................................... OK
test:mechanical-small-problem.frictionless-nodal-lm-mortar-disp ........................................... OK
test:verification/patch_tests/plane_2.frictionless_kin .................................................... OK
test:verification/patch_tests/ring_4.glued_pen ............................................................ OK
test:verification/patch_tests/cyl_4.frictionless_kin ...................................................... OK
test:verification/patch_tests/plane_4.glued_pen ........................................................... OK
test:verification/patch_tests/ring_2.frictionless_kin ..................................................... OK
test:verification/patch_tests/brick_2.glued_pen ........................................................... OK
test:verification/patch_tests/cyl_3.frictionless_kin ...................................................... OK
test:verification/patch_tests/single_pnt_2d.frictionless_kin .............................................. OK
test:verification/patch_tests/plane_3.frictionless_kin .................................................... OK
test:verification/patch_tests/brick_4.glued_kin ........................................................... OK
test:verification/patch_tests/ring_3.frictionless_kin ..................................................... OK
test:verification/patch_tests/cyl_1.frictionless_kin ...................................................... OK
test:nodal_area.3D2Proc ...................................................................... [min_cpus=2] OK
test:fieldsplit.mortar-FS ................................................................................. OK
test:verification/patch_tests/brick_1.frictionless_kin .................................................... OK
test:verification/patch_tests/plane_1.frictionless_pen .................................................... OK
test:verification/patch_tests/ring_1.frictionless_pen ..................................................... OK
test:verification/patch_tests/plane_2.frictionless_pen .................................................... OK
test:verification/patch_tests/cyl_2.frictionless_pen ...................................................... OK
test:verification/patch_tests/ring_2.frictionless_pen ..................................................... OK
test:verification/patch_tests/ring_4.frictionless_kin ..................................................... OK
test:verification/patch_tests/plane_4.frictionless_kin .................................................... OK
test:verification/patch_tests/cyl_3.frictionless_pen ...................................................... OK
test:verification/patch_tests/single_pnt_2d.frictionless_pen .............................................. OK
test:mechanical-small-problem.frictionless-nodal-lm-mortar-disp-action .................................... OK
test:nodal_area.2D ........................................................................................ OK
test:verification/patch_tests/brick_2.frictionless_kin .................................................... OK
test:verification/patch_tests/cyl_1.frictionless_pen ...................................................... OK
test:normalized_penalty.normalized_penalty_Q8 ............................................................. OK
test:verification/patch_tests/brick_1.frictionless_pen .................................................... OK
test:verification/patch_tests/single_pnt_2d.frictionless_pen_contact_line_search .......................... OK
test:incremental_slip.slip ................................................................................ OK
test:tension_release.4ElemTensionRelease ..................................................... [min_cpus=4] OK
test:verification/patch_tests/brick_2.frictionless_pen .................................................... OK
test:bouncing-block-contact.frictionless-mortar-min-lm-mortar-disp-coarse ................................. OK
test:normalized_penalty.normalized_penalty_kin ............................................................ OK
test:dual_mortar.dm_contact .................................................................. FAILED (ERRMSG)
test:ranfs-and-scaling.auto-scaling .......................................................... FAILED (ERRMSG)
test:kinematic-and-scaling.auto-scaling ...................................................... FAILED (ERRMSG)
test:verification/overclosure_removal.test ................................................... FAILED (ERRMSG)
test:verification/patch_tests/automatic_patch_update.sliding_blocks .......................... FAILED (ERRMSG)
test:non-singular-frictional-mortar.nonsingular .............................................. FAILED (ERRMSG)
test:mechanical_constraint.frictionless_kinematic ............................................ FAILED (ERRMSG)
test:bouncing-block-contact.frictionless-nodal-min-lm-mortar-disp-coarse ..................... FAILED (ERRMSG)
test:tan-pen-and-scaling.auto-scaling ........................................................ FAILED (ERRMSG)
test:frictional/sliding_elastic_blocks_2d.sliding_elastic_blocks_2d .......................... FAILED (ERRMSG)
test:adaptivity.contact_initial_adaptivity ................................................... FAILED (ERRMSG)
test:dual_mortar.std_contact ................................................................. FAILED (ERRMSG)
test:catch_release.catch_release ............................................................. FAILED (ERRMSG)
test:frictional/single_point_2d.single_point_2d .............................................. FAILED (ERRMSG)
test:fieldsplit.2block-3D .................................................................... FAILED (ERRMSG)
test:ranfs-and-scaling.no-scaling ............................................................ FAILED (ERRMSG)
test:kinematic-and-scaling.no-scaling ........................................................ FAILED (ERRMSG)
test:bouncing-block-contact.frictionless-nodal-fb-lm-mortar-disp-coarse ...................... FAILED (ERRMSG)
test:tan-pen-and-scaling.no-scaling .......................................................... FAILED (ERRMSG)
test:frictional/sliding_elastic_blocks_2d.sliding_elastic_blocks_2d_tp ....................... FAILED (ERRMSG)
test:frictional/single_point_2d.single_point_2d_predictor .................................... FAILED (ERRMSG)
test:mechanical_constraint.frictionless_penalty .............................................. FAILED (ERRMSG)
test:verification/patch_tests/brick_3.frictionless_pen ....................................... FAILED (ERRMSG)
test:bouncing-block-contact.frictionless-nodal-min-lm-nodal-disp-coarse ...................... FAILED (ERRMSG)
test:pressure.pressureAugLag ................................................................. FAILED (ERRMSG)
test:frictional/single_point_2d.single_point_2d_tp ........................................... FAILED (ERRMSG)
test:verification/patch_tests/cyl_4.frictionless_pen ......................................... FAILED (ERRMSG)
test:verification/patch_tests/plane_3.frictionless_pen ....................................... FAILED (ERRMSG)
test:bouncing-block-contact.frictionless-nodal-fb-lm-nodal-disp-coarse ....................... FAILED (ERRMSG)
test:verification/patch_tests/ring_3.frictionless_pen ........................................ FAILED (ERRMSG)
test:verification/patch_tests/ring_1.frictionless_aug ........................................ FAILED (ERRMSG)
test:verification/patch_tests/brick_3.frictionless_aug ....................................... FAILED (ERRMSG)
test:verification/patch_tests/plane_1.frictionless_aug ....................................... FAILED (ERRMSG)
test:verification/patch_tests/plane_2.frictionless_aug ....................................... FAILED (ERRMSG)
test:verification/patch_tests/cyl_2.frictionless_aug ......................................... FAILED (ERRMSG)
test:verification/patch_tests/ring_2.frictionless_aug ........................................ FAILED (ERRMSG)
test:verification/patch_tests/cyl_4.frictionless_aug ......................................... FAILED (ERRMSG)
test:verification/patch_tests/ring_4.frictionless_pen ........................................ FAILED (ERRMSG)
test:verification/patch_tests/cyl_3.frictionless_aug ......................................... FAILED (ERRMSG)
test:verification/patch_tests/plane_4.frictionless_pen ....................................... FAILED (ERRMSG)
test:bouncing-block-contact/grid-sequencing.grid_sequencing .................................. FAILED (ERRMSG)
test:verification/patch_tests/plane_3.frictionless_aug ....................................... FAILED (ERRMSG)
test:verification/patch_tests/ring_3.frictionless_aug ........................................ FAILED (ERRMSG)
test:verification/patch_tests/cyl_1.frictionless_aug ......................................... FAILED (ERRMSG)
test:verification/patch_tests/ring_1.mu_0_2_aug .............................................. FAILED (ERRMSG)
test:verification/patch_tests/plane_1.mu_0_2_aug ............................................. FAILED (ERRMSG)
test:verification/patch_tests/plane_2.mu_0_2_aug ............................................. FAILED (ERRMSG)
test:verification/patch_tests/cyl_2.mu_0_2_aug ............................................... FAILED (ERRMSG)
test:verification/patch_tests/ring_2.mu_0_2_aug .............................................. FAILED (ERRMSG)
test:verification/patch_tests/cyl_4.mu_0_2_pen ............................................... FAILED (ERRMSG)
```

**Comment by lindsayad** [comment]
Also for thermal contact:
```
test:homogenization.heatConduction_test ...................................................... FAILED (ERRMSG)
test:heat_conduction/2d_quadrature_gap_heat_transfer.moving .................................. FAILED (ERRMSG)
test:heat_conduction/3d_quadrature_gap_heat_transfer.moving .................................. FAILED (ERRMSG)
--------------------------------------------------------------------------------------------------------------
```
Interesting about the homogenization failure. Are periodic bcs maybe stateful?

**Comment by lindsayad** [comment]
It turns out that MOOSE is not the right place to do this because we apply constraints in libmesh after completing `FEProblemBase::computeResidualSys`. That is the reason for all the failures observable on the most recent round of testing related to adaptivity and periodic boundary conditions.

File: framework/src/utils/PetscSupport.C [hunk 1]
```diff
@@ -927,6 +933,22 @@ colorAdjacencyMatrix(PetscScalar * adjacency_matrix,
...
+
+  petsc_options.flags.erase("-snes_converged_reason");
+}
+
+void
+disableLinearConvergedReason(FEProblemBase & fe_problem)
+{
+  auto & petsc_options = fe_problem.getPetscOptions();
+
+  petsc_options.flags.erase("-ksp_converged_reason");
```
Comment (fdkong) [inline framework/src/utils/PetscSupport.C:949]: Could you remind me what reason we can not print out the colorful converged reason?

Can we set up our own reason view?

**Review by fdkong — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
Per my slack comment (
https://moosedevelopers.slack.com/archives/D0103R1901Z/p1594397950015900),
I don't think you can set a custom convergence reason viewer. You can
obviously set a custom convergence check, but I don't think you can set a
custom reason viewer.

**Comment by lindsayad** [comment]
You're the PETSc expert. I'll defer to your knowledge.

**Comment by lindsayad** [comment]
I guess I should bring back the linear convergence check again, and we can just print the reason from the convergence check. Right now in the nonlinear convergence check we generate a message that we put into `PetscInfo`. That doesn't seem very useful, e.g. our users will never look at that. We could just do `_console` output I'm assuming from `FEProblemBase::checkLinearConvergence` (if I brought it back). And then we could change the messaging in `FEProblemBase::checkNonlinearConvergence` to go to `_console` as well instead of to `PetscInfo`. But then we will be printing a Moose convergence reason instead of a PETSc one, which is perhaps what we want?

**Comment by fdkong** [comment]
Looks good for me. I do not have requirements on color things.

**Review by fdkong — APPROVED** [review]

**Comment by fdkong** [comment]
Bringing  LinearChecker back is too much

**Pull request merged**